FuelLabs/fuel-indexer
# chore: introduce ephemeral DBs for e2e tests

## Changelog
- Add `FromStr` implementation for `DatabaseConfig`
- Add `DbUrlParseError` to `IndexerConfigError`
- Add `TestPostgresDb` fixture
-- `new()`: Uses the `DATABASE_URL` environment variable to connect to the `postgres` user on a Postgres server. It then generates a 10-character string and creates a database with that name
-- `teardown()`: Reconnects to the Postgres server, terminates all connections to the unique database, and then drops said database
- Add custom `Drop` implementation for `TestPostgresDb`
-- `drop()` calls `teardown()` by spawning a blocking thread as destructors cannot be run asynchronously
- Adjusts service and API server fixtures to take an optional DB connection string; if absent, the default database values will be used, just as before.
- Adjust Postgres E2E and integration tests to use `TestPostgresDb`, which isolates the data for each test

## Testing Plan
CI should pass as it uses the same `DATABASE_URL` environment variable that is checked in the DB construction.

I repeatedly ran the E2E and integration tests and ensured that all testing databases are in fact ephemeral. You can test this by doing the following (if you have Postgres installed through homebrew):
1. `psql -l` - this will list all of the databases on the server
```
❯ psql -l
                             List of databases
   Name    |  Owner   | Encoding | Collate | Ctype |   Access privileges
-----------+----------+----------+---------+-------+-----------------------
 postgres  | deekerno | UTF8     | C       | C     |
 template0 | deekerno | UTF8     | C       | C     | =c/deekerno          +
           |          |          |         |       | deekerno=CTc/deekerno
 template1 | deekerno | UTF8     | C       | C     | =c/deekerno          +
           |          |          |         |       | deekerno=CTc/deekerno
(3 rows)
```
2. In your terminal session (or in your shell config), make sure that the `DATABASE_URL` environment variable is set to use a user that has the `CREATEDB` permission. For example, my Postgres server is on my local host and default port and has a `postgres` superuser with `my-secret` as a password.
```
export DATABASE_URL="postgres://postgres:my-secret@localhost:5432"
```
3. Run the E2E tests: `cargo test --features e2e,postgres e2e -- --test-threads=1`. They should pass. Feel free to run the integration tests as well: `cargo test --features e2e,postgres integration -- --test-threads=1`
4. Run `psql -l` again; there should be no extra databases and the output should be the same as the original.

## Notes
Eventually, I'd like to do one more adjustment and spawn the indexer service/API server through `lazy_static!`. With some adjustments to the `Drop` behaviors of those structs and changes to the tests themselves, that would allow for us to run the entire workspace test suite in parallel without having to exclude `fuel-indexer-tests` for single threading (that means one less CI job). But again, _eventually_.
